scalding-core/src/main/scala/com/twitter/scalding/ExecutionOptimizationRules.scala

```diff
@@ -147,10 +151,23 @@ object ExecutionOptimizationRules {
       def apply(tup: (A, B, C, ConcurrentExecutionContext)): Future[(D, E)] =
         (Execution.failFastZip(fn1(tup), fn2(tup))(tup._4))
     }
+
+    case class MapWriteFn[A, B, C, D, E](
+      fn1: ((A, B, C, ConcurrentExecutionContext)) => Future[D],
+      mapFn: D => E) extends Function1[(A, B, C, ConcurrentExecutionContext), Future[E]] {
+
+      def apply(tup: (A, B, C, ConcurrentExecutionContext)): Future[E] = fn1(tup).map(mapFn)(tup._4)
+    }
+
     def mergeWrite[A, B](w1: WriteExecution[A], w2: WriteExecution[B]): WriteExecution[(A, B)] = {
       val newFn = ComposeWriteFn(w1.result, w2.result)
       WriteExecution(w1.head, w1.tail ::: (w2.head :: w2.tail), newFn)
     }
+    def mapWrite[A, B](w: WriteExecution[A], fn: A => B): WriteExecution[B] = {
+      WriteExecution(w.head, w.tail, MapWriteFn(w.result, fn))
+    }
+
+    // either NonWrite(ex: Execution), Write(ex: Execution), composed: NonWrite, Write
```

Comment:
This comment seems stray. Can we move it closer to where we want it?

Reply:
Removed